File: core/ZigzagConversion/ZigzagConversion.py
def convert(s: str, numRows: int) -> str:
    res_matrix = [[] for _ in range(numRows)]
    cur_point, flag = 0, "down"
    for ch in s:
        res_matrix[cur_point].append(ch)
        if flag == "down" and cur_point + 1 < numRows:
            cur_point += 1
        elif flag == "down":
            cur_point -= 1
            flag = "up"
        elif flag == "up" and cur_point - 1 >= 0:
            cur_point -= 1
        else:
            cur_point += 1
            flag = "down"

    return "".join(["".join(item) for item in res_matrix])


if __name__ == '__main__':
    print(convert(s="AB", numRows=1))

File: core/ZigzagConversion/ZigzagConversion.java
package com.dbc;

import java.util.ArrayList;
import java.util.List;

public class ZigzagConversion {
    public static String convert(String s, int numRows) {
        List<Character>[] resMatrix = new List[numRows];
        for (int i = 0; i < numRows; i++) {
            resMatrix[i] = new ArrayList<>();
        }
        int curPoint = 0, flag = 0;
        for (int i = 0; i < s.length(); i++) {
            resMatrix[curPoint].add(s.charAt(i));
            if (flag == 0 && curPoint + 1 < numRows) {
                curPoint++;
            } else if (flag == 0 && curPoint - 1 >= 0) {
                curPoint--;
                flag = 1;
            } else if (curPoint - 1 >= 0) {
                curPoint--;
            } else if (curPoint + 1 < numRows) {
                curPoint++;
                flag = 0;
            }
        }

        StringBuffer res = new StringBuffer();
        for (List<Character> chList : resMatrix) {
            for (Character ch : chList) {
                res.append(ch);
            }
        }
        return res.toString();
    }
}
